nifi-nar-bundles/nifi-jmx-bundle/nifi-jmx-processors/src/main/java/org/apache/nifi/processors/jmx/GetJMX.java

```diff
@@ -0,0 +1,406 @@
+
+    private List<PropertyDescriptor> descriptors;
+
+    private Set<Relationship> relationships;
+    private final BlockingQueue<Record> queue = new LinkedBlockingQueue<>();
+
+    private final Set<Record> inProcess = new HashSet<>();    // guarded by queueLock
+    private final Set<Record> recentlyProcessed = new HashSet<>();    // guarded by queueLock
+
+    private final Lock queueLock = new ReentrantLock();
+    private final Lock listingLock = new ReentrantLock();
+    private final AtomicLong queueLastUpdated = new AtomicLong(0L);
+
+    private ListFilter whiteListFilter;
+    private ListFilter blackListFilter;
+
+    private final ComponentLog logger = getLogger();
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptors = new ArrayList<PropertyDescriptor>();
+        descriptors.add(HOSTNAME);
+        descriptors.add(PORT);
+        descriptors.add(WHITELIST);
+        descriptors.add(BLACKLIST);
+        descriptors.add(POLLING_INTERVAL);
+        descriptors.add(BATCH_SIZE);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<Relationship>();
+        relationships.add(REL_SUCCESS);
+        this.relationships = Collections.unmodifiableSet(relationships);
+        whiteListFilter = new ListFilter( "" );
+        blackListFilter = new ListFilter( "" );
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        queue.clear();
+    }
+
+    private Set<Record> performQuery( ProcessContext context) throws ProcessException {
+        PropertyValue hostname = context.getProperty(HOSTNAME);
+        PropertyValue port = context.getProperty(PORT);
+        PropertyValue whitelist = context.getProperty(WHITELIST);
+        PropertyValue blacklist = context.getProperty(BLACKLIST);
+
+        Set<Record> metricsSet = new HashSet<Record>();
+
+        whiteListFilter.setListString(whitelist.getValue());
+        blackListFilter.setListString( blacklist.getValue() );
+
+        try {
+            HashMap<String,Object> env = new HashMap<String,Object>();
+
+            StringBuilder urlStr = new StringBuilder("service:jmx:rmi:///jndi/rmi://");
+            urlStr.append(hostname).append(":");
+            urlStr.append(port);
+            urlStr.append("/jmxrmi");
+
+            JMXServiceURL url = new JMXServiceURL(urlStr.toString());
+            JMXConnector jmxc = JMXConnectorFactory.connect(url, null);
+
+            MBeanServerConnection mbsc = jmxc.getMBeanServerConnection();
+
+            Set<ObjectName> mbeans = mbsc.queryNames(null, null);
```

Comment:
Instead of querying everything and then using whitelist and blacklist, wouldn't be easier to directly specify what we want when executing the query? I'd assume that users would either want everything or just a subset of known MBean names. That would simplify a lot the processor.